### PR TITLE
[TACHYON-794] Added test coverage for inodeTree

### PR DESCRIPTION
**Pull request by yupeng9**
https://tachyon.atlassian.net/browse/TACHYON-794

### CODE REVIEW

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeDirectory.java [hunk 1]
```diff
@@ -34,16 +34,19 @@
...
     }
   };
-  private IndexedSet.FieldIndex mNameIndex = new IndexedSet.FieldIndex<Inode>() {
+  private IndexedSet.FieldIndex<Inode> mNameIndex = new IndexedSet.FieldIndex<Inode>() {
+    @Override
     public Object getFieldValue(Inode o) {
       return o.getName();
     }
   };
+  @SuppressWarnings("unchecked")
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeDirectory.java]: why is this necessary?

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java [hunk 2]
```diff
@@ -38,16 +38,18 @@
...
   private final InodeDirectory mRoot;
 
-  private final IndexedSet.FieldIndex mIdIndex = new IndexedSet.FieldIndex<Inode>() {
+  private final IndexedSet.FieldIndex<Inode> mIdIndex = new IndexedSet.FieldIndex<Inode>() {
+    @Override
     public Object getFieldValue(Inode o) {
       return o.getId();
     }
   };
+  @SuppressWarnings("unchecked")
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java]: why is this necessary?

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+import tachyon.master.next.block.BlockMaster;
+import tachyon.thrift.BlockInfoException;
+import tachyon.thrift.FileAlreadyExistException;
+import tachyon.thrift.FileDoesNotExistException;
+import tachyon.thrift.InvalidPathException;
+
+/**
+ * Unit tests for InodeTree.
+ */
+public final class InodeTreeTests {
```
Comment (gpang) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: I think the convention is to name it "InodeTreeTest"

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  @Before
+  public void before() {
+    mTree = new InodeTree(new BlockMaster());
+  }
+
+  @Test
+  public void createPathTest() throws Exception {
+    // create directory
+    mTree.createPath(TEST_URI, Constants.KB, false, true);
+    Inode test = mTree.getInodeByPath(TEST_URI);
```
Comment (gpang) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: Can you also test that this is a InodeDirectory?

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    Assert.assertEquals(TEST_PATH, test.getName());
+
+    // create nested directory
+    mTree.createPath(NESTED_URI, Constants.KB, true, true);
+    Inode nested = mTree.getInodeByPath(NESTED_URI);
```
Comment (gpang) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: Can you test this type?

Also, are there no tests for creating a file with createPath() (and testing its type)?

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+  @Test
+  public void createRootPathTest() throws Exception {
+    mThrown.expect(FileAlreadyExistException.class);
+    mThrown.expectMessage("/");
+
+    mTree.createPath(new TachyonURI("/"), Constants.KB, false, true);
+  }
+
+  @Test
+  public void createEmptyFileTest() throws Exception {
```
Comment (gpang) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: I think this should be called createFileInvalidBlockSizeTest or something similar, since it is about invalid block sizes.

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    Assert.assertEquals(2, inodes.size());
+    // delete the nested inode
+    mTree.deleteInode(nested);
+    inodes = mTree.getInodeChildrenRecursive((InodeDirectory) mTree.getInodeById(0));
+    // only /nested left
+    Assert.assertEquals(1, inodes.size());
+  }
+
+  @Test
+  public void deleteNoexistingInodeTest() throws Exception {
```
Comment (gpang) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: "Noexisting" -> "Nonexisting"

**Comment by gpang** [comment]
Thanks for these tests! I left some comments, and could you resolve the conflicts?

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeDirectory.java (hunk 1 above)
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeDirectory.java]: To suppress the warning: A generic array of IndexedSet.FieldIndex<Inode> is created for 
 a varargs parameter

File: servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java (hunk 2 above)
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/next/filesystem/meta/InodeTree.java]: ditto

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java (hunk 3 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: I actually feel adding the suffix of Tests might be better, because this class contains several tests, no?

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java (hunk 4 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: done

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java (hunk 5 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: Good point. Added

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java (hunk 6 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: done

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java (hunk 7 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: done

**Comment by yupeng9** [comment]
Thanks for the comments, @gpang. Please take a look at the new changes.

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java (hunk 3 above)
Comment (gpang) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: I think the current convention should be kept for now, to stay consistent. If it is changed later, all of the test file names can be changed at the same time.

**Comment by gpang** [comment]
I think the test name should stick with the current convention. Also, could you resolve merge conflicts? Thanks!

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java (hunk 3 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTests.java]: Sounds good to me.

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+/**
+ * Unit tests for InodeTree.
+ */
+public final class InodeTreeTest {
+  private static final String TEST_PATH = "test";
+  private static final TachyonURI TEST_URI = new TachyonURI("/test");
+  private static final TachyonURI NESTED_URI = new TachyonURI("/nested/test");
+  private static final TachyonURI NESTED_FILE_URI = new TachyonURI("/nested/test/file");
+  private InodeTree mTree;
```
Comment (uronce-cc) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java]: Format: an empty line above and below needed.

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+    mTree.createPath(new TachyonURI("/"), Constants.KB, false, true);
+  }
+
+  @Test
+  public void createFileWithInvalidBlockSizeTest() throws Exception {
+    mThrown.expect(BlockInfoException.class);
+    mThrown.expectMessage("Invalid block size 0");
+
+    mTree.createPath(TEST_URI, 0, false, false);
```
Comment (uronce-cc) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java]: Could add more tests for block size < 0

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    mTree.createPath(NESTED_URI, Constants.KB, true, false);
+    mTree.createPath(new TachyonURI("/nested/test/test"), Constants.KB, true, false);
+  }
+
+  @Test
+  public void getInodeByNonexistingPathTest() throws Exception {
+    mThrown.expect(InvalidPathException.class);
+    mThrown.expectMessage("Could not find path: /test");
+
+    mTree.getInodeByPath(TEST_URI);
```
Comment (uronce-cc) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java]: Could be better to test the case when part of the path is non existing.

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    mThrown.expect(FileDoesNotExistException.class);
+    mThrown.expectMessage("Inode id 1 does not exist");
+
+    Inode testFile = new InodeFile("testFile1", 1, 1, Constants.KB, System.currentTimeMillis());
+    mTree.deleteInode(testFile);
+  }
+
+  @Test
+  public void setPinnedTest() throws Exception {
+    Inode nested = mTree.createPath(NESTED_URI, Constants.KB, true, true);
```
Comment (uronce-cc) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java]: What about putting these `mTree.createPath` into helper functions like `createNestedUri()`, `createNestedFileUri()`...  they could be helper for future tests because determining each parameter in `createPath` seems not intuitive, especially the `true/false`.

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java (hunk 8 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java]: done

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java (hunk 9 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java]: done

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java (hunk 10 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java]: good point

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java (hunk 11 above)
Comment (yupeng9) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeTreeTest.java]: We could, but it wouldn't save too much code. In fact, the function calls in this class are with different parameters, including the block size and the boolean ones. What do you think?

File: servers/src/test/java/tachyon/master/next/filesystem/meta/InodeDirectoryTest.java [hunk 12]
```diff
@@ -36,8 +36,8 @@ public void addChildrenTest() {
     InodeFile inodeFile2 = createInodeFile(3);
     inodeDirectory.addChild(inodeFile1);
     inodeDirectory.addChild(inodeFile2);
-    Assert.assertEquals(createBlockId(2), (long) inodeDirectory.getChildrenIds().get(0));
-    Assert.assertEquals(createBlockId(3), (long) inodeDirectory.getChildrenIds().get(1));
+    Assert.assertTrue(inodeDirectory.getChildrenIds().contains(createBlockId(2)));
+    Assert.assertTrue(inodeDirectory.getChildrenIds().contains(createBlockId(3)));
```
Comment (gpang) [inline servers/src/test/java/tachyon/master/next/filesystem/meta/InodeDirectoryTest.java]: Maybe you should save the set of children ids, and then compare. Also, you should assert the length == 2.

**Comment by gpang** [comment]
Just a minor comment, but in general, looks good!

**Comment by uronce-cc** [comment]
There are merge conflicts, but LGTM.

**Comment by gpang** [comment]
Please resolve the merge conflicts, but looks good!

**Comment by gpang** [comment]
LGTM
